src/BaseControllerV2.ts

```diff
@@ -43,20 +51,63 @@ export type IsJsonable<T> =
       never;
 
 /**
- * Controller class that provides state management and subscriptions
+ * An function to derive state.
+ *
+ * This function will accept one piece of the controller state (one property),
+ * and will return some derivation of that state.
+ *
+ * @param value - A piece of controller state
+ * @returns Something derived from controller state
+ */
+export type StateDeriver<T> = (value: IsJsonable<T>) => IsJsonable<Json>;
+
+/**
+ * State metadata.
+ *
+ * This metadata describes which parts of state should be persisted, and how to
+ * get an anonymized representation of the state.
+ */
+export type StateMetadata<T> = {
+  [P in keyof T]: StatePropertyMetadata<T[P]>;
+};
+
+/**
+ * Metadata for a single state property
+ *
+ * @property persist - Indicates whether this property should be persisted
+ *   (`true` for persistent, `false` for transient), or is set to a function
+ *   that derives the persistent state from the state.
+ * @property anonymous - Indicates whether this property is already anonymous,
+ *   (`true` for anonymous, `false` if it has potential to be personally
+ *   identifiable), or is set to a function that returns an anonymized
+ *   representation of this state.
+ */
+export interface StatePropertyMetadata<T> {
+  persist: boolean | StateDeriver<T>;
+  anonymous: boolean | StateDeriver<T>;
+}
+
+type Json = null | boolean | number | string | Json[] | { [prop: string]: Json } | Partial<Record<never, never>>;
```

Comment:
I don't think the `Partial` is required around that `Record<never, never>` :thinking: I'll double-check this.

That was meant to cover the empty object case, as I didn't know how to adjust `{ [prop: string]: Json }` to allow zero keys.

Reply:
So the goal with that last type is to throw an error if none of the other cases match first, right?. I think you can just wrap your object with Partial instead of the Record.

Reply:
or, you could add ` | {}` to explicitly allow a empty object.

Reply:
I tried testing this type with the same TypeScript playground test suite used to validate the IsJsonable type, and discovered that an empty object is already accepted because of the ` { [prop: string]: Json }` entry.

https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgzgewHZQLxSQVwDbagHygCMEFsIBDFQrAWyIgCcCo5hGBLJAcxfmQDaAXRYBvKALCMEYAFyt2XbkPn8UAXwDcAWABQegGaYkAY2AdkUYBDYAeACoA+ABQArRElUeAlPIBuCBwAJlCielBQAPSRUPY2wFAccIkowAAW0CYItGAc2BTmyHrqenpc1owGFCbQAIKh4VAU8nQMjDq6Jfq6JvlwyQBCDboRFACELZj0TB1dZUgVVTVQAGrDowD8k9PtxaW65UxL0AAa603yACIFELP7h5XV0ACa5xRbUNfWd90Px1AALTe2zaLGMQQgBi4ECCP3miyeUAA6m8Ps5vGhHE0kCA4T1kGxsSA0BgcHgKMkqLj9tEoFkcnkmFAAPIAaT01jYziwuG8HU5wGcVWwcAgfI58WcAAZxboBc4AEQK2Xy4QqyWidTqrkCKVCbWC8TNKBSqBa-mSqlNZK1A3OK0UqADO1G+TiIjydiYCAAGjpbvU6j9QXkAk1foATH6FWkOAqhOaJTrw6FjTG40HSbg-QBmfUdPS0+m5cjMdLSADucCTgqQEArnxu6LtAjrDa+EHR+ZrzldGHrjes6LNdodyRWo5x1qgJ0nxMdzzn04BS8dSLt4Mh0KC3h7AFZp60mYRN1C6zuOkA

Though weirdly, without `{}` or `Record` included, two of the test cases are showing as invalid because of the error "Index signature is missing". Also it looks like this last entry in the type union was wholly responsible for allowing Date objects and other non-JSON things. :thinking:

Reply:
Oh wow, I think `IsJsonable` is wrong. It doesn't check for key types that aren't supported by JSON. That error is legitimate.

Reply:
I'll clean this up in a later PR. There are a few changes needed to these types I think.

Reply:
This was cleaned up here: #373